topology/probes/k8sctl.go

```diff
@@ -0,0 +1,488 @@
+	rv := reflect.ValueOf(v)
+	if rv.IsNil() {
+		return "nil"
+	}
+	pv := reflect.Indirect(rv).Interface()
+	return fmt.Sprintf("*%v", pv)
+}
+
+func (k8s *K8SCtlProbe) updatek8sNP() {
+	t := config.GetConfig().GetDuration("k8s.polling_rate")
+	logging.GetLogger().Infof("Update every %s", t)
+	for {
+		k8s.k8sNPOLD = make([]k8sNP, k8s.maxNP)
+		k8s.k8sNPOLD = k8s.k8sNPNEW
+		k8s.k8sNPNEW = make([]k8sNP, k8s.maxNP)
+
+		k8s.getk8sPods()
+		k8s.syncWithK8s()
+		k8s.getk8sNP()
+		k8s.copyEdgeIds()
+		k8s.updateNPLinks()
+		time.Sleep(t)
+	}
+}
+
+// Copy corresponding Edge ID to each Link (Edge ID is assigned to k8s.k8sNPNEW by k8s.updateNPLinks() only when the Edge is first added
+// Need to preserve it when updating k8s.k8sNPOLD/NEW ...
+func (k8s *K8SCtlProbe) copyEdgeIds() {
+	for i, link := range k8s.k8sNPNEW {
+		if link.from == "" || link.to == "" {
+			break
+		}
+		edge := k8s.findEdge(link)
+		if edge != "" {
+			k8s.k8sNPNEW[i].e = edge
+		}
+	}
+}
+
+// retrun Edge ID of NP link is exist
+func (k8s *K8SCtlProbe) findEdge(link k8sNP) graph.Identifier {
+	for _, linkOLD := range k8s.k8sNPOLD {
+		if linkOLD.from == link.from && linkOLD.to == link.to {
+			return linkOLD.e
+		}
+	}
+	return ""
+}
+
+// retrun the skydive ID of the matched pod according to given namespace
+func (k8s *K8SCtlProbe) findPodByNS(ns string) (bool, []k8sPodEdge) {
+	var ret []k8sPodEdge
+	var valid = false
+	for name, data := range k8s.k8sPods {
+		if data.Namespace == ns {
+			var pod k8sPodEdge
+			pod.GraphNode = data.GraphNode
+			pod.Name = name
+			ret = append(ret, pod)
+			valid = true
+		}
+	}
+	return valid, ret
+}
+
+// retrun the skydive ID of the matched pod according to given label key/value
+func (k8s *K8SCtlProbe) findPodByLabel(key string, value string, ns string) (bool, k8sPodEdge) {
+	var ret k8sPodEdge
+	for name, data := range k8s.k8sPods {
+		labels := data.Labels
+		if labels[key] == value {
+			if ns != "" {
+				if data.Namespace == ns {
+					ret.GraphNode = data.GraphNode
+					ret.Name = name
+					return true, ret
+				}
+			} else {
+				ret.GraphNode = data.GraphNode
+				ret.Name = name
+				return true, ret
+			}
+		}
+	}
+	return false, ret
+}
+
+// sync between pods (k8s.k8sPods) as reported by k8s and skydive netns IDs (k8s.NetnsIDs)
+func (k8s *K8SCtlProbe) syncWithK8s() {
+	for _, netns := range k8s.Graph.GetNodes(graph.Metadata{"Type": "netns"}) {
+		for name, data := range k8s.k8sPods {
+			netnsName, err := netns.GetFieldString("Name")
+			if err != nil {
+				logging.GetLogger().Debugf("Can't find Name field of netns ID %s", netns.ID)
+				continue
+			}
+			if strings.Contains(netnsName, string(data.UID)) {
+				// update k8sPods
+				data.GraphNode = netns.ID
+				k8s.k8sPods[name] = data
+				logging.GetLogger().Debugf("Update pod %s with ID %s", name, data.GraphNode)
+			}
+		}
+	}
+}
+
+// gets pods from k8s
+func (k8s *K8SCtlProbe) getk8sPods() bool {
+	pods, err := k8s.clientset.Core().Pods("").List(v1.ListOptions{})
```

Comment:
do we need to get here pods while we should already have everything needed in the Graph.Using graph.Lookup* functions with the correct metadata should be enough to find them instead of using findByLabel/ByNs ?